quickwit-actors/src/actor_handle.rs

```diff
@@ -0,0 +1,165 @@
+        let actor_instance_name = ctx.actor_instance_name().to_string();
+        let ctx_clone = ctx.clone();
+        tokio::task::spawn(async move {
+            interval.tick().await;
+            while ctx.kill_switch().is_alive() {
+                interval.tick().await;
+                if !ctx.progress().harvest_changes() {
+                    if ctx.get_state() == ActorState::Terminated {
+                        return;
+                    }
+                    error!(actor=%actor_instance_name, "actor-timeout");
+                    ctx.kill_switch().kill();
+                    return;
+                }
+            }
+        });
+        ActorHandle {
+            join_handle,
+            last_state,
+            actor_context: ctx_clone,
+        }
+    }
+
+    /// Process all of the pending message, and returns a snapshot of
+    /// the observable state of the actor after this.
+    ///
+    /// This method is mostly useful for tests.
+    ///
+    /// To actually observe the state of an actor for ops purpose,
+    /// prefer using the `.observe()` method.
+    ///
+    /// This method timeout if reaching the end of the message takes more than an HEARTBEAT.
+    /// Hence, it is only useful or unit tests.
+    pub async fn process_pending_and_observe(&self) -> Observation<A::ObservableState> {
+        let (tx, rx) = oneshot::channel();
+        if self
+            .actor_context
+            .mailbox()
+            .send_actor_message(ActorMessage::Observe(tx))
+            .await
+            .is_err()
+        {
+            error!("Failed to send message");
+        }
+        // TODO The timeout is required here. If the actor fails, the inbox is properly dropped but the send channel might actually
+        // prevent the onechannel Receiver from being dropped.
+        let observable_state_res = tokio::time::timeout(crate::HEARTBEAT, rx).await;
+        let state = self.last_state.borrow().clone();
+        match observable_state_res {
+            Ok(Ok(_)) => Observation::Running(state),
+            Ok(Err(_)) => Observation::Terminated(state),
+            Err(_) => Observation::Timeout(state),
+        }
+    }
+
+    /// Terminates the actor, regardless of whether there are pending messages or not.
+    pub async fn finish(&self) {
+        let (tx, rx) = oneshot::channel();
+        let _ = self
+            .actor_context
+            .mailbox()
+            .send_command(Command::Stop(tx))
+            .await;
+        let _ = rx.await;
+    }
+
+    pub async fn join(self) -> Result<(ActorTermination, A::ObservableState), JoinError> {
+        let termination = self.join_handle.await?;
+        let observation = self.last_state.borrow().clone();
+        Ok((termination, observation))
+    }
+
+    /// Observe the current state.
+    ///
+    /// If a message is currently being processed, the observation will be
+    /// after its processing has finished.
```

Comment:
/// The observation will be scheduled as a command message, therefore it will be executed
/// after the current active message and the current command queue have been processed.